File: dsl-api/src/main/java/cbs/dsl/api/TriggerDefinition.java
package cbs.dsl.api;

import lombok.Getter;

import java.time.Instant;

/**
 * Defines a trigger for scheduling mass operations.
 *
 * <p>This is an abstract type with concrete subclasses for different trigger kinds.
 */
public abstract class TriggerDefinition {

  private TriggerDefinition() {}

  /** A cron-expression-based trigger. */
  @Getter
  public static final class CronTrigger extends TriggerDefinition {
    private final String expression;

    public CronTrigger(String expression) {
      this.expression = expression;
    }
  }

  /** A one-time trigger at a specific instant. */
  @Getter
  public static final class OnceTrigger extends TriggerDefinition {
    private final Instant at;

    public OnceTrigger(Instant at) {
      this.at = at;
    }
  }

  /** A periodic trigger defined by days, hours, and minutes. */
  @Getter
  public static final class EveryTrigger extends TriggerDefinition {
    private final int days;
    private final int hours;
    private final int minutes;

    public EveryTrigger(int days, int hours, int minutes) {
      this.days = days;
      this.hours = hours;
      this.minutes = minutes;
    }
  }

  /** A signal-based trigger. */
  @Getter
  public static final class SignalTrigger extends TriggerDefinition {
    private final SignalTypes.Signal signal;

    public SignalTrigger(SignalTypes.Signal signal) {
      this.signal = signal;
    }
  }
}

File: dsl-api/src/main/kotlin/cbs/dsl/api/TriggerDefinition.kt
package cbs.dsl.api

sealed class TriggerDefinition {
  data class CronTrigger(val expression: String) : TriggerDefinition()

  data class OnceTrigger(val at: java.time.Instant) : TriggerDefinition()

  data class EveryTrigger(val days: Int = 0, val hours: Int = 0, val minutes: Int = 0) :
      TriggerDefinition()

  data class SignalTrigger(val signal: SignalTypes.Signal) : TriggerDefinition()
}
